### PR TITLE
[Core] Simplify Raylet Client

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

This PR removes duplicated code in Raylet Client. This also helps improve code reuse for later refactoring of plasma store.

This PR also removes all platform-specific code in Raylet Client.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/common/client_connection.cc
Comment (mehrdadn) [inline src/ray/common/client_connection.cc:61]: Do you know if this function can be removed without much negative impact? It doesn't really seem to provide a useful abstraction.

**Review by mehrdadn — APPROVED** [review]
Left a couple minor notes but LGTM assuming tests pass.

File: src/ray/common/client_connection.cc
Comment (suquark) [reply src/ray/common/client_connection.cc:61]: Do you mean "ConnectSocketRetry"? I find it ("ConnectSocketRetry") will be used later in 3 more places (it is originally used in Plasma Store). This is why I try to keep this function.

Comment (mehrdadn) [reply src/ray/common/client_connection.cc:61]: Sorry no, I meant `ServerConnection::Create`. I realize you haven't modified it, I just mentioned it since it seems like it might be worth the cleanup after the surrounding refactor.

File: src/ray/raylet/raylet_client.cc
Comment (suquark) [reply src/ray/raylet/raylet_client.cc:99]: There are 4 points for me to use `std::vector` instead.

1. I find it is not convenient to use `std::unique_ptr<uint8_t[]>` in general cases with `boost::asio`. `boost::asio` has native support for `std::vector` as a buffer. To use `std::unique_ptr<uint8_t[]>` with `boost::asio`, we need to use `get()` to extract the underlying raw pointer, and pass the length of the buffer as an independent parameter. In all other places, we have replaced `std::unique_ptr<uint8_t[]>` with `std::vector`.

2. For our general message processing, we need a buffer with a length for the purpose of various data parsing (flatbuffer, msgpack, protobuf, etc). For me `std::vector` is a better abstraction with some boundary guarantee.

3. The performance is not a problem, because we need to allocating memory for `std::unique_ptr<uint8_t[]>` dynamically too, and we use `reserve` for `std::vector`.

4. Google C++ style prefers pointer type for `output arguments`. This creates ugly types like `std::unique_ptr<uint8_t[]>*`. I really don't like it. Why not just use managed buffers like `std::vector` to control the lifetime? C++ should ensure zero-cost abstraction.

File: src/ray/common/client_connection.cc
Comment (suquark) [reply src/ray/common/client_connection.cc:61]: Oh, this magic `ClientConnection::Create` is critical to maintain the lifetime of this object (it is extremely confusing, but I remember it is from boost official examples). One example is here:

https://github.com/ray-project/ray/blob/fcdf410ae1bb5071e7d92174eace7d79be2d4ef9/src/ray/raylet/raylet.cc#L129-L148

In this example, it looks like `new_connection` will go out of scope, and we are leaking memory. But actually `ClientConnection::Create` calls `client_handler`, and `client_handler` will call `ClientConnection::ProcessMessages()`, then `ClientConnection::ProcessMessages()` will general a `shared_ptr` of itself via `shared_ClientConnection_from_this()` in a boot::asio callback, so that its lifetime will be only controlled by `boost::asio`. This is quite tricky, but there is hardly other ways to achieve it.

Comment (suquark) [reply src/ray/common/client_connection.cc:61]: I agree with you that forcing `ServerConnection` to be a `shared_ptr` could be over decorating, but I don't have a strong mind to remove it.

File: src/ray/raylet/raylet_client.cc
Comment (mehrdadn) [reply src/ray/raylet/raylet_client.cc:99]: 1. [Asio's primary support](https://www.boost.org/doc/libs/1_73_0/doc/html/boost_asio/overview/core/buffers.html) is for e.g. [`mutable_buffer`](https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/reference/buffer.html) which is a pointer and a size, not vector.

2. I don't think we need `size()` though, do we? We weren't using it so far (given `std::unique_ptr` sufficed for our usage), and even after the change, I see
    ```
    flatbuffers::GetRoot<protocol::RegisterClientReply>(reply.data());
    ```
    which still ignores the size. I might be missing something though.

3. Vector has additional initialization overhead (I'm not referring to allocation) compared to a plain dynamic array.

4. I agree `std::unique_ptr<uint8_t[]> *` isn't pretty, but `std::vector<uint8_t> *` is the same thing with size/capacity added. The double-nullity is nice to avoid, but you can avoid that with a reference rather than a pointer. In fact, the Google style guide (which I don't really defer to personally) [says the same thing](https://google.github.io/styleguide/cppguide.html): _"required (non-nullable) output and input/output parameters should usually be references"_.

Personally my biggest concern though is that `vector` is semantically meant to be a random-access container and not a buffer of raw memory, so it ends up being somewhat misleading IMO. But that said, I don't feel too strongly about it, so if you disagree despite these feel free to keep it.

File: src/ray/common/client_connection.cc
Comment (mehrdadn) [reply src/ray/common/client_connection.cc:61]: Sorry, I was just referring to `ServerConnection::Create`, not `ClientConnection::Create`. Does removing `ServerConnection::Create` have any impact on users of `ClientConnection::Create`? I was imagining it would only imply a trivial change to callers callers to use `std::make_shared`, but I'm not sure.

Comment (mehrdadn) [reply src/ray/common/client_connection.cc:61]: (In any case I'm imagining this is a < 5-minute change, so just disregard my comment if it's more complicated.)

Comment (suquark) [reply src/ray/common/client_connection.cc:61]: I can do it in a later PR.

File: src/ray/raylet/raylet_client.cc
Comment (suquark) [reply src/ray/raylet/raylet_client.cc:99]: The size is used everywhere in Plasma Store, which is very painful if we only keep `std::unique_ptr<uint8_t[]> *`. You can see the `MessageHandler` in ClientConnection has been using `std::vector`.

https://github.com/ray-project/ray/blob/fcdf410ae1bb5071e7d92174eace7d79be2d4ef9/src/ray/object_manager/plasma/store.cc#L894-L916

We need to use size here:

https://github.com/ray-project/ray/blob/8c985dc79764408be7004f327a6c0f5a625b2271/src/ray/object_manager/plasma/protocol.h#L36-L40

**Comment by suquark** [comment]
The CI errors seem not related and also appear in the upstream. I'll merge this PR.

**Pull request merged**
